Add collector for definitions

HaltCatchFire opened the pull request.

ruoso commented:
Naming-wise, I don't think you need "Collector" in the name of the class, since it's already in the `collectors` namespace, at which point we can not abbreviate.

We probably should include in the name some information about the fact that it's collecting them and doing a "normalization" process.

HaltCatchFire commented:
@ruoso 

I removed the abbreviations; I'd previously abbreviated because "References" was abbreviated in "VariableRefs", so I thought "Defs" would be more consistent than "Definitions".

I agree that "Collector" doesn't make sense in the class name, so I removed that as well.

The code no longer does the normalization (that'll be done in an external tool), but it now maps "FileID"s to the "NamedDecl"s that were matched, so it might be a good idea to give that field a more descriptive name.

ruoso commented:
This needs rebasing from master

ruoso commented on include/clangmetatool/collectors/definitions.h:
```diff
@@ -0,0 +1,26 @@
+#ifndef INCLUDED_DEFINITIONS_H
+#define INCLUDED_DEFINITIONS_H
```
The guard should have the whole path to the file, not just the basename.

ruoso commented on include/clangmetatool/collectors/definitions_data.h:
```diff
@@ -0,0 +1,26 @@
+class DefinitionsData {
+    public:
+    // filename that contains definition -> AST Node
+    std::unordered_multimap<const clang::FileID, const clang::NamedDecl *, FileIdHasher> defs;
```
In the `IncludeGraph` we actually use the `UID` of the file, we even have `clangmetatool::types::FileUID` that we use for that. It's a more friendly thing to use as the key of the map since it is just a number.
